DESIGN.md

```diff
@@ -0,0 +1,773 @@
+internally. It is possible to have two operations in-flight from a single
+resource associated with different queues and threads.
+
+Because operations are not `Send`, tasks awaiting these operations are also not
+`Send`, making them unable to be spawned using the spawn function from the Tokio
+crate. The tokio-uring crate will provide a spawn function that accepts not
+`Send` tasks.
+
+When using multiple isolated runtime threads, balancing load between them
+becomes challenging. Applications must take care to ensure load remains balanced
+across threads, and strategies tend to vary. For example, a TCP server can
+distribute accepted connections across multiple threads, ideally while
+maintaining equal load across threads. One approach is to submit "accept"
+operations for the same listener concurrently across all the runtime threads
+while ensuring overloaded workers back-off. Defining runtime load is out of this
+crate's scope and left to the application, though pseudocode follows.
+
+```rust
+let listener = Arc::new(TcpListener::bind(...));
+
+spawn_on_each_thread(async {
+    loop {
+        // Wait for this worker to have capacity. This
+        // ensures there are a minimum number of workers
+        // in the runtime that are flagged as with capacity
+        // to avoid total starvation.
+        current_worker::wait_for_capacity().await;
+        
+        let socket = listener.accept().await;
+        
+        spawn(async move { ... });
+    }
+})
+```
+
+## Reading and writing
+
+Read and write operations require passing ownership of buffers to the kernel.
+When the operation completes, the kernel returns ownership of the buffer to the
+caller. The caller is responsible for allocating the buffer's memory and
+ensuring it remains alive until the operation completes. Additionally, while the
+kernel owns the memory, the process may not read from or write to the buffer. By
+designing the Rust APIs using ownership passing, Rust enforces the requirements
+at compile time. The following example demonstrates reading and writing with a
+file resource.
+
+```rust
+use tokio_uring::buf;
+
+/// The result of an operation that includes a buffer. The buffer must
+/// be returned to the caller when the operation completes successfully
+/// or fails.
+///
+/// This is implemented as a new type to implement std::ops::Try once
+/// the trait is stabilized.
+type BufResult<T> = (std::io::Result<T>, buf::Slice);
+
+// Positional read and write function definitions
+impl File {
+    async fn read_at(&self, buf: buf::Slice, pos: u64) -> BufResult<usize>;
+    
+    async fn write_at(&self, buf: buf::Slice, pos: u64) -> BufResult<usize>;
+}
+
+/// The caller allocates a buffer
+let buf = buf::IoBuf::with_capacity(4096);
+
+// Read the first 1024 bytes of the file, when `std::ops::Try` is
+// stable, the `?` can be applied directly on the `BufResult` type.
+let BufResult(res, buf) = file.read_at(0, buf.slice(0..1024)).await;
```

Comment:
I have a question on this. What if a future is canceled without being polled to complete? In this case,
* The kernel is still handling the request.
* The slice is dropped (as part of the future). (Not sure about this, point me out if I'm wrong)
* When kernel completes the request, memory reference of this slice might be no longer valid, which causes a segment fault.

Reply:
I've got the answer in this proposal. Thanks for your work!

> In this case, tokio-uring will close the resource in the background, avoiding blocking the runtime. 

But it seems that this part has not been implemented yet. Current implementation might block the runtime by using a spin loop. https://github.com/tokio-rs/tokio-uring/blob/master/src/driver/op.rs#L210